SCRUM 18 added like feature schema, route, and controller

alcarrillomunoz opened the pull request:
**Description**
This merge adds the like feature to like/unlike an item, and get the like count and user data that liked each item.

**Changes Made**
- created Like schema that gets assigned an item and user
- created like route and controller for liking/unliking an item, and getting like data from an item (like count and users that liked the item)

**Additional Notes**
Testing was done via Postman. Error handling set to check for user, item, and like data. 

**Merge Request Checklists**
Code follows project coding guidelines.
Documentation reflects the changes made.

orangesoftware reviewed (approved):
For future:
API rest usually follow next pattern using versioning:
/v1/api/likes        -> GET Method to get all the elements,
/v1/api/likes/:id   -> GET Method to get 1 element
/v1/api/likes        -> POST Method to add a new element
When you are using a post method try to avoid passing a URL parameter, some servers are not accepting that.

Miminiverse commented on src/controllers/like.js:
Since we already have a middleware authenticateUser to check User, we don't need to check User again, we can just use it directly. 
These are redundant in both likeItem and getLike
const user = await User.findOne({ _id: req.user._id });
    if (!user) {
      return res
        .status(StatusCodes.NOT_FOUND)
        .json(createResponse("error", "User not found"));
    }

Miminiverse replied:
When unliking, you return the like object that was found before deletion. We should only return the ID of the deleted like or just a success message without the old object.

else {
      await Like.findOneAndDelete({ _id: like._id });
      res.status(StatusCodes.OK).json(
        createResponse("success", "Item was unliked", {
          like: like,
        })
      );
    }

Miminiverse replied:
Finding zero likes is a successful result, not an error or not found condition. An item could have no like which is totally normal.

if (!likeData || likeCount === 0) {
      res
        .status(StatusCodes.NOT_FOUND)
        .json(createResponse("error", "Item like data not found"));
    }

Miminiverse commented on src/models/Like.js:
This is good enough for now. 
Just be mindful with this current setup, a user could technically like an item multiple times if logic fails. In the future we can modify this to prevent a user from liking the same item multiple times.

alcarrillomunoz commented:
Changes made:
- removed URL paramater to POST, sending item ID in body now
- added validation for itemId provided 
- removed check for user exists to eliminate redundancy 
- deleted like response no longer returning deleted like data
- removed error return for likeCount = 0 (thank you, I was questioning this!)

Miminiverse reviewed (approved):
Great job building the Like feature! I'm excited to see this in the app soon. 
I still think this should not throw error at all, because if no likes are found, it's still correct and should not be identified as error. If you still want to check this, maybe change the condition to if (likeCount === 0) or if (likeData.length === 0), return 200 and a message saying there are no likes 
 if (!likeData) {
      res
        .status(StatusCodes.NOT_FOUND)
        .json(createResponse("error", "Item like data not found"));
    } else

alcarrillomunoz commented:
Oh gotcha @Miminiverse this was removed!
